Framework/DataHandling/src/LoadEventNexusIndexSetup.cpp

```diff
@@ -0,0 +1,158 @@
+#include "MantidTypes/SpectrumDefinition.h"
+
+using namespace Mantid::API;
+using namespace Mantid::Indexing;
+
+namespace Mantid {
+namespace DataHandling {
+
+LoadEventNexusIndexSetup::LoadEventNexusIndexSetup(
+    MatrixWorkspace_const_sptr instrumentWorkspace, const int32_t min,
+    const int32_t max, const std::vector<int32_t> range)
+    : m_instrumentWorkspace(instrumentWorkspace), m_min(min), m_max(max),
+      m_range(range) {}
+
+std::pair<int32_t, int32_t> LoadEventNexusIndexSetup::eventIDLimits() const {
+  return {m_min, m_max};
+}
+
+IndexInfo LoadEventNexusIndexSetup::makeIndexInfo() {
+  // The default 1:1 will suffice but exclude the monitors as they are always in
+  // a separate workspace
+  auto detIDs = m_instrumentWorkspace->getInstrument()->getDetectorIDs(true);
+  const auto &detectorInfo = m_instrumentWorkspace->detectorInfo();
+  std::vector<SpectrumDefinition> specDefs;
+  for (const auto detID : detIDs)
+    specDefs.emplace_back(detectorInfo.indexOf(detID));
+  // We need to filter based on detector IDs, but use IndexInfo for filtering
+  // for a unified filtering mechanism. Thus we set detector IDs as (temporary)
+  // spectrum numbers.
+  IndexInfo indexInfo(
+      std::vector<SpectrumNumber>(detIDs.begin(), detIDs.end()));
+  indexInfo.setSpectrumDefinitions(specDefs);
+
+  auto filtered = filterIndexInfo(indexInfo);
+
+  // If there is no actual filter, spectrum numbers are contiguous and start at
+  // 1. Otherwise spectrum numbers are detector IDs. This is legacy behavior
+  // adopted from EventWorkspaceCollection.
+  if (filtered.size() == indexInfo.size())
+    filtered.setSpectrumNumbers(1, static_cast<int32_t>(filtered.size()));
+
+  return filtered;
+}
+
+IndexInfo LoadEventNexusIndexSetup::makeIndexInfo(
+    const std::vector<std::string> &bankNames) {
+  const auto &componentInfo = m_instrumentWorkspace->componentInfo();
+  std::vector<SpectrumDefinition> spectrumDefinitions;
+  const auto &instrument = m_instrumentWorkspace->getInstrument();
+  for (const auto &bankName : bankNames) {
+    const auto &bank = instrument->getComponentByName(bankName);
+    std::vector<size_t> dets;
+    if (bank) {
+      const auto bankIndex = componentInfo.indexOf(bank->getComponentID());
+      dets = componentInfo.detectorsInSubtree(bankIndex);
+      for (const auto detIndex : dets)
+        spectrumDefinitions.emplace_back(detIndex);
+    }
+    if (dets.empty())
+      throw std::runtime_error("Could not find the bank named '" + bankName +
+                               "' as a component assembly in the instrument "
+                               "tree; or it did not contain any detectors. Try "
+                               "unchecking SingleBankPixelsOnly.");
+  }
+  Indexing::IndexInfo indexInfo(spectrumDefinitions.size());
+  indexInfo.setSpectrumDefinitions(std::move(spectrumDefinitions));
+  return indexInfo;
+}
+
+IndexInfo LoadEventNexusIndexSetup::makeIndexInfo(
```

Comment:
I can see you have tests for this, but is it worth adding comments to the effect of 2,3,4 under "Currently not fixed" in the PR description. Again this is only to aid future refactoring/fixing.

Reply:
Yes, will add doxygen once we know if this is actually the correct thing to do, or if we maybe want to fix even more things.